src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -712,16 +712,46 @@ private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSy
         internal MethodSymbol TryFindDisposePatternMethod(TypeSymbol exprType, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
         {
             LookupResult lookupResult = LookupResult.GetInstance();
-            
-            MethodSymbol disposeMethod = FindPatternMethod(exprType, WellKnownMemberNames.DisposeMethodName, lookupResult, syntaxNode, warningsOnly: true, diagnostics, syntaxNode.SyntaxTree, MessageID.IDS_Disposable);
-            lookupResult.Free();
 
-            if (disposeMethod?.ReturnsVoid == false)
+            DiagnosticBag initialDiagnostics = DiagnosticBag.GetInstance();
+            MethodSymbol disposeMethod = FindPatternMethodRelaxed(exprType,
+                                                                  WellKnownMemberNames.DisposeMethodName,
+                                                                  lookupResult,
+                                                                  syntaxNode,
+                                                                  warningsOnly: true,
+                                                                  initialDiagnostics,
+                                                                  syntaxNode.SyntaxTree,
+                                                                  MessageID.IDS_Disposable);
+
+            if (disposeMethod != null && disposeMethod.ReturnsVoid == false)
             {
-                diagnostics.Add(ErrorCode.WRN_PatternBadSignature, syntaxNode.Location, exprType, MessageID.IDS_Disposable.Localize(), disposeMethod);
+                ReportPatternWarning(initialDiagnostics, exprType, disposeMethod, syntaxNode, MessageID.IDS_Disposable);
                 disposeMethod = null;
             }
 
+            if (disposeMethod is null)
+            {
+                lookupResult.Clear();
+                disposeMethod = FindPatternExtensionMethod(exprType,
+                                                           WellKnownMemberNames.DisposeMethodName,
+                                                           lookupResult,
+                                                           syntaxNode,
+                                                           diagnostics,
+                                                           MessageID.IDS_Disposable);
```

Comment:
Can the pattern case be handled before checking `ReturnsVoid`, so there is only one call to `ReportPatternWarning`?